### PR TITLE
Switch to sqrt(precision) representation in Gaussian

### LINKED ISSUE #567: Switch to sqrt(prescision) representation in Gaussian?
Addresses #559 

This issue proposes switching the `Gaussian` parameters from `(info_vec, precision)` to `(info_vec, prec_sqrt)`, following @fehiepsi's work in https://github.com/pyro-ppl/pyro/pull/2019.

## Motivation

Our original motivation for representing Gaussians as `(info_vec, precision)` was to support operations on rank-deficient precision matrices, which occur in conditional probability distributions. This design choice allows us to uniformly handle priors, conditional distributions, and likelihoods by treating all three agnostically as mere factors in a factor graph.

However while the `(info_vec, precision)` representation is numerically stable, it is computationally inefficient when making low-dimensional observations of a high-dimensional variable.  For example to store a conditional distribution of a 1-dimensional variable given a 1000-dimensional variable, the precision matrix has 1001**2 elements, but since it has rank 1 its square root would cost only 1001 elements.  Indeed we recognized this https://github.com/pyro-ppl/pyro/pull/2005 and https://github.com/pyro-ppl/funsor/pull/217 and created a special `AffineNormal` pattern to avoid materializing rank-1 precision matrices.

An alternative representation is the classic square root information filter (SRIF), explored by @fehiepsi in https://github.com/pyro-ppl/pyro/pull/2019. This represents a Gaussian as a pair `(info_vec, prec_sqrt)`, of shapes `batch_shape + (dim,)` and `batch_shape + (dim, rank)` respectively, so that
```py
precision = prec_sqrt @ prec_sqrt.transpose(-2, -1)
```
Advantages of the square root information representation include:
- space storage cost is `O(dim * rank)` which can be much smaller than `O(dim ** 2)`
- ops.add can be implemented by mere concatenation (optionally [followed by compression](https://github.com/pyro-ppl/pyro/pull/2019/files#r317249339) if `rank > dim` or maybe `rank > 1.5 * dim`)
- we could drop `AffineNormal`

## Design questions

- [x] Should we make this change?
- [x] If we change `Gaussian` interface what will break?
    - the [arXiv paper](https://arxiv.org/abs/1910.10775) will be out of date
    - I believe all Pyro & NumPyro code uses `to_funsor()`, so will be unaffected
- [x] How should we make this change?  Possibilities include:
    - :+1: Simply replace the existing `Gaussian`.
    - ~~Build a separate `GaussianS`.~~
    - ~~Create a new `ConditionalGaussian` as the basic thing and make `Gaussian` a mere alias~~
        (but how do we unify priors, conditionals, and likelihoods?)
    - ~~Create an abstraction for structured precision matrices as in GPyTorch.~~
        (or is it cleaner to allow structured square root matrices, since they have natural block structure?)

### PR DESCRIPTION
**Pull request by fritzo**
Resolves #567 
Adapts @fehiepsi's https://github.com/pyro-ppl/pyro/pull/2019

This switches the internal Gaussian representation to a numerically stable and space efficient representation
```diff
- Gaussian(info_vec, precision, inputs)
+ Gaussian(white_vec, prec_sqrt, inputs)
```
In the new parametrization, Gaussians represent the log-density function
```py
Gaussian(white_vec, prec_sqrt, OrderedDict(x=Reals[n]))
  = -1/2 || x @ prec_sqrt - white_vec ||^2
```
These two parameters are shaped to efficiently represent low-rank data:
```py
assert white_vec.shape == batch_shape + (rank,)
assert prec_sqrt.shape == batch_shape + (dim, rank)
```
reducing space complexity from `O(dim(dim+1))` to `O(rank(dim+1))`. In my real-world example rank=1, dim=2369, and batch_shape=(1343,), so the space reduction is 30GB → 13MB.

Computations is cheap in this representation: addition amounts to concatenation, and plate-reduction amounts to transpose and reshape.  Some ops are only supported on full-rank Gaussians, and I've added checks based on the new property `.is_full_rank`.  This partial support is ok because the Gaussian funsors that arise in Bayesian models are all full rank due to priors (notwithstanding numerical loss of rank).

Because the Gaussian funsor is internal, the interface change should not cause breakage of most user code, since most user code uses `to_funsor()` and `to_data()` with backend-specific distributions.  One broken piece of user code is Pyro's [AutoGaussianFunsor](https://github.com/pyro-ppl/pyro/blob/ac3c588c7e18a695000f87322e037d31ad2900e5/pyro/infer/autoguide/gaussian.py#L434-L436) which will need an update (and which will be sped up).

As suggested by @eb8680, I've added some optional kwarg parametrizations and properties to support conversion to other Gaussian representations, e.g. `g = Gaussian(mean=..., covariance=..., inputs=...)` and `g._mean`, `g._covariance`.  This allows more Gaussian math to live in gaussian.py.

## Tested
- [x] added new tests
- [x] pass existing funsor tests (linear algebra)
- [x] pass existing funsor tests (patterns)
- [x] pass NumPyro tests (conjugacy)
- [x] pass Pyro tests on a branch (https://github.com/pyro-ppl/pyro/pull/2943)
- [x] check computational cost on the pyro-cov model (results: under 12GB memory)

### CODE REVIEW

**Comment by fritzo** [comment]
@eb8680 could we pair code on Gaussian patterns this week?  I think this PR now has correct linear algebra, but it produces slightly different patterns.  Specifically, the new square root representation can no longer be negated, so we'll need to keep `Unary(ops.neg, Gaussian)` lazy.

**Comment by fritzo** [comment]
This is currently failing a Pyro AutoGaussian pyro-cov test, where the optimized Gaussian is
<details>

```
Contraction(ops.null, ops.add,
 frozenset(),
 (Contraction(ops.logaddexp, ops.add,
   frozenset({Variable('coef_scale__BOUND_38', Real)}),
   (Gaussian(
     torch.tensor(...1..., dtype=torch.float64),
     torch.tensor(...1 x 1..., dtype=torch.float64),
     (('coef_scale__BOUND_38', Real),)),
    Contraction(ops.logaddexp, ops.add,
     frozenset({Variable('rate_scale__BOUND_37', Real)}),
     (Gaussian(
       torch.tensor(...1..., dtype=torch.float64),
       torch.tensor(...1 x 1..., dtype=torch.float64),
       (('rate_scale__BOUND_37', Real),)),
      Contraction(ops.logaddexp, ops.add,
       frozenset({Variable('coef__BOUND_39', Reals[6])}),
       (Gaussian(
         torch.tensor(...6..., dtype=torch.float64),
         torch.tensor(...7 x 6..., dtype=torch.float64),
         (('coef__BOUND_39', Reals[6]),
          ('coef_scale__BOUND_38', Real),)),
        Contraction(ops.add, ops.null,
         frozenset({Variable('place__BOUND_33', Bint[4])}),
         (Contraction(ops.logaddexp, ops.null,
           frozenset({Variable('rate__BOUND_32', Reals[5])}),
           (Gaussian(
             torch.tensor(...4 x 5..., dtype=torch.float64),
             torch.tensor(...4 x 12 x 5..., dtype=torch.float64),
             (('place__BOUND_33', Bint[4]),
              ('rate__BOUND_32', Reals[5]),
              ('rate_scale__BOUND_37', Real),
              ('coef__BOUND_39', Reals[6]),)),)),)),)),)),)),
  Contraction(ops.logaddexp, ops.add,
   frozenset({Variable('init_loc_scale__BOUND_36', Real)}),
   (Gaussian(
     torch.tensor(...1..., dtype=torch.float64),
     torch.tensor(...1 x 1..., dtype=torch.float64),
     (('init_loc_scale__BOUND_36', Real),)),
    Contraction(ops.logaddexp, ops.add,
     frozenset({Variable('init_scale__BOUND_34', Real)}),
     (Gaussian(
       torch.tensor(...1..., dtype=torch.float64),
       torch.tensor(...1 x 1..., dtype=torch.float64),
       (('init_scale__BOUND_34', Real),)),
      Contraction(ops.logaddexp, ops.add,
       frozenset({Variable('init_loc__BOUND_35', Reals[5])}),
       (Gaussian(
         torch.tensor(...5..., dtype=torch.float64),
         torch.tensor(...6 x 5..., dtype=torch.float64),
         (('init_loc__BOUND_35', Reals[5]),
          ('init_loc_scale__BOUND_36', Real),)),
        Contraction(ops.add, ops.null,
         frozenset({Variable('place__BOUND_31', Bint[4])}),
         (Contraction(ops.logaddexp, ops.null,
           frozenset({Variable('init__BOUND_30', Reals[5])}),
           (Gaussian(
             torch.tensor(...4 x 5..., dtype=torch.float64),
             torch.tensor(...4 x 11 x 5..., dtype=torch.float64),
             (('place__BOUND_31', Bint[4]),
              ('init__BOUND_30', Reals[5]),
              ('init_scale__BOUND_34', Real),
              ('init_loc__BOUND_35', Reals[5]),)),)),)),)),)),)),))
```

</details>
and backward sampling erroneously tries to call `._sample({"init_loc"}, {}` on this Gaussian:
```
Gaussian(
 torch.tensor(...5..., dtype=torch.float64),
 torch.tensor(...6 x 5..., dtype=torch.float64),
 (('init_loc', Reals[5]),
  ('init_loc_scale__BOUND_36', Real),))
```

**Comment by fritzo** [comment]
@eb8680 could you please review this PR in general?
@fehiepsi could you please review linear algebra?

I'm happy to walk you through the changes over zoom.

**Comment by fehiepsi** [comment]
Whoa, impressive work to make this possible! I haven't looked into the code yet but my general concern would be on marginalization and compress rank logics. I will look into the details later of this week.

**Comment by fritzo** [comment]
Thanks for reviewing @eb8680 and @fehiepsi!  I believe I've addressed all comments.

I plan to add more patterns in subsequent PRs that handle Gaussian variable elimination, e.g. in #553 and https://github.com/pyro-ppl/funsor/tree/tractable-for-gaussians
